### PR TITLE
Image undistortion for fisheye lenses

### PR DESCRIPTION
**Pull request by guidoAI**
I have added undistortion functions to the vision library. These can be used to convert image coordinates from distorted fisheye lenses to undistorted coordinates and back, taking into account the camera calibration matrix and the distortion of the specific lens. The distortion function is the one from Dhane, P., Kutty, K., & Bangadkar, S. (2012). A generic non-linear method for fisheye correction. International Journal of Computer Applications, 51(10). (invertible fisheye model), as it is an invertible model. 

I also made a module, undistort_image, that uses the functions to undistort a whole image. This is quite a slow process, so it is not really advisable to use this in a vision pipeline leading to control. However, it can be used to find the right undistortion parameter k, and shows that the undistortion functions work.

I have tested the functions and module on a Bebop 2, using the front image. I added a specific airframe to the TU Delft airframes for this purpose.

### CODE REVIEW

File: sw/airborne/modules/computer_vision/undistort_image.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+float min_x_normalized;
+float max_x_normalized;
+float dhane_k;
+float center_ratio;
+
+struct video_listener *listener = NULL;
+
+// Camera calibration matrix (of the Bebop2 for a specific part of the visual sensor):
+static float K[9] = {189.69f, 0.0f, 165.04f,
```
Comment (podhrmic) [inline sw/airborne/modules/computer_vision/undistort_image.c]: this should be made configurable as it is camera/drone specific

File: sw/airborne/modules/computer_vision/undistort_image.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                     0.0f, 188.60f, 118.44f,
+                     0.0f, 0.0f, 1.0f};
+
+//static float k;
```
Comment (podhrmic) [inline sw/airborne/modules/computer_vision/undistort_image.c]: remove unnecessary comment

File: sw/airborne/modules/computer_vision/undistort_image.c [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+// Function
+struct image_t *undistort_image_func(struct image_t *img);
```
Comment (podhrmic) [inline sw/airborne/modules/computer_vision/undistort_image.c]: declaration vs definition - probably no need to declare the function, unless you want to put this in a separate header

File: sw/airborne/modules/computer_vision/undistort_image.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  float normalized_step = (max_x_normalized - min_x_normalized) / img->w;
+  float h_w_ratio = img->h / (float) img->w;
+  float min_y_normalized = h_w_ratio * min_x_normalized;
+  float max_y_normalized = h_w_ratio * max_x_normalized;
+
+  // create an image of the same size:
+  struct image_t img_distorted;
+  image_create(&img_distorted, img->w, img->h, img->type);
+  // this info is not created in image_create, but is necessary for streaming...:
+  //img_distorted.ts = img->ts;
```
Comment (podhrmic) [inline sw/airborne/modules/computer_vision/undistort_image.c]: so should it be here?

File: sw/airborne/modules/computer_vision/undistort_image.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  // fill the image again, now with the undistorted image:
+  float x_pd, y_pd;
+  uint32_t x_pd_ind, y_pd_ind;
+  uint32_t x = 0;
+  for(float x_n = min_x_normalized; x_n < max_x_normalized; x_n += normalized_step, x++) {
+    uint32_t y = 0;
+    for(float y_n = min_y_normalized; y_n < max_y_normalized; y_n += normalized_step, y++) {
+      // for faster execution but smaller FOV, uncomment the next line:
```
Comment (podhrmic) [inline sw/airborne/modules/computer_vision/undistort_image.c]: this should be controllable via a flag or a variable that the user can switch on/off

**Comment by guidoAI** [comment]
As described in the newest commit, cdf55d6, I followed up on all advice, adding camera calibration parameters and removing unnecessary comments.

**Comment by OpenUAS** [comment]
Was talking about the "Documentation" of the XML module, it had this incorrect line. No worries, I honord the pull and will fixup the documentation self.  BTW I also have only "limited time™" . :wink:

**Comment by guidoAI** [comment]
Thanks for looking at it and merging it. 

On my mobile phone, the description tags did not show, so I only saw that "the colorfilter" was incorrect... 

You can use this as a description:
Undistortion module. It takes a fisheye image and undistorts it with the help of a camera calibration matrix and distortion parameters, as given by the user.

**Pull request merged**